### ISSUE TITLE
create a default user for android devices

### ISSUE THREAD

**Issue by leomaxi** [open-learning-exchange/planet#1277]
create a default user for android devices with permission to 
read databases, write to databases with exception of _users

**Comment by lmmrssa** [comment]
@paulbert we will require satellite user to be system admin. We will not require normal user of it.
@leomaxi needs to fetch user salt and other details on tablet to make them able to login.

**Comment by paulbert** [comment]
@leomaxi @lmmrssa @dogi This is not a secure way to handle the situation, as it would place a very easy to crack full administrator on our nations & communities.

There needs to be a very good reason to change this to a system admin, otherwise we should remove this admin from the dev and use a regular user.

In addition, whether this is a regular user or an admin it is not necessary to fetch the salt or hashed password.  He can use the plain text password to authenticate.

**Comment by leomaxi** [comment]
@paulbert , Can you explain how i can use plain text password to authenticate?. How would the tablet know the users password is correct. The tablet application doesn't do live - online, connected to server authentication but rather downloads user records from the server so when the server is off, random members / users can login on the tablet and use materials offline. So at the time this authentication will be happening, the server is / in most cases will be offline / off. 
The user we need is a system admin user that can be used to query the database for all users (encrypted) and saved in realm database. 

- If we hardcode a user in the app, only persons working on the app will know the user authentication protocol

- If we have a username and password sit on the server (Generated on planet installation). And is saved in a Json or directory where the tablet can read and use this details to communicate with the server as an admin. {each installation will have a unique username and password}

**Comment by paulbert** [comment]
The _users database is public to read, and you can create users (which will not have write access, but will still be stored on the DB) without admin credentials.

The plan was to give you one user with write access to the necessary databases and you would use those credentials whenever the tablet is connected to send data to CouchDB.  The administrator will get the password (pin) for that user from the browser app management dashboard and that should be saved to the tablets when they first open the app so any time they connect to CouchDB the app has the pin available.

When putting data on CouchDB you can authenticate either by creating a session by POSTing the username and pin to the `_security` endpoint [like this](http://docs.couchdb.org/en/2.1.1/intro/security.html#cookie-authentication) or by adding the user name & pin to the front of the request.

**Comment by lmmrssa** [comment]
I think when tablet is first setup or when tablet needs to fetch users then they can request it using the admin credentials (username/password) to planet. And all other sync will be done by satellite on later phases so that tablet does not require to have admin credentials always. Only when you want users to be synced then will have to be done by system admin.
@paulbert @leomaxi does that sound like a plan?

**Comment by paulbert** [comment]
@lmmrssa We can fetch user details without credentials.  That plan would be OK as it does not involve creating an easily hacked admin, but there is no need to use admin credentials so we do not need to complicate our solution like that.

**Comment by leomaxi** [comment]
To sync databases with CouchDbClient, user / tablet has have admin rights.. Currently satellite doesn't have admin right and therefore crushes the app.
See below... 
 
```
   --------- beginning of system
08-09 14:17:12.536 3128-3166/org.ole.planet.takeaway D/EGL_emulation: eglMakeCurrent: 0xa34d5dc0: ver 2 0
08-09 14:17:12.598 3128-3166/org.ole.planet.takeaway D/EGL_emulation: eglMakeCurrent: 0xa34d5dc0: ver 2 0
08-09 14:17:12.613 3128-3166/org.ole.planet.takeaway D/EGL_emulation: eglMakeCurrent: 0xa34d5dc0: ver 2 0
08-09 14:17:12.699 3128-3237/org.ole.planet.takeaway I/CouchDbClient: > GET/
08-09 14:17:12.703 3128-3237/org.ole.planet.takeaway I/CouchDbClient: < Status: 200
08-09 14:17:12.720 3128-3237/org.ole.planet.takeaway I/CouchDbClient: > GET/_users/_all_docs?include_docs=true
08-09 14:17:12.724 3128-3237/org.ole.planet.takeaway I/CouchDbClient: < Status: 401
    
    --------- beginning of crash
08-09 14:17:12.731 3128-3237/org.ole.planet.takeaway E/AndroidRuntime: FATAL EXCEPTION: Thread-177
    Process: org.ole.planet.takeaway, PID: 3128
    org.lightcouch.CouchDbException: Unauthorized{"error":"unauthorized","reason":"You are not a server admin."}
    
        at org.lightcouch.CouchDbClientBase.validate(CouchDbClientBase.java:659)
        at org.lightcouch.CouchDbClientAndroid$2.process(CouchDbClientAndroid.java:193)
        at org.apache.http.protocol.BasicHttpProcessor.process(BasicHttpProcessor.java:309)
        at org.apache.http.protocol.HttpRequestExecutor.postProcess(HttpRequestExecutor.java:324)
        at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:469)
        at org.apache.http.impl.client.AbstractHttpClient.execute(AbstractHttpClient.java:560)
        at org.lightcouch.CouchDbClientBase.executeRequest(CouchDbClientBase.java:480)
        at org.lightcouch.CouchDbClientBase.get(CouchDbClientBase.java:531)
        at org.lightcouch.CouchDbClientBase.get(CouchDbClientBase.java:542)
        at org.lightcouch.View.queryForStream(View.java:130)
        at org.lightcouch.View.query(View.java:142)
        at org.ole.planet.takeout.service.TransactionSyncManager$1.execute(TransactionSyncManager.java:27)
        at io.realm.Realm.executeTransaction(Realm.java:1405)
        at org.ole.planet.takeout.service.TransactionSyncManager.syncDb(TransactionSyncManager.java:23)
        at org.ole.planet.takeout.service.SyncManager$1.run(SyncManager.java:86)
        at java.lang.Thread.run(Thread.java:818)
08-09 14:17:12.763 3128-3166/org.ole.planet.takeaway D/EGL_emulation: eglMakeCurrent: 0xa34d5dc0: ver 2 0
08-09 14:17:13.220 3128-3128/org.ole.planet.takeaway D/OLE: log: Destroyed 
```

**Comment by paulbert** [comment]
@lmmrssa @dogi @leomaxi
Possible solution to allow access to credentials:

We can create a `tablet_user` database which can only be read by admins & the satellite user and when setting up for the first time create a continuous replication for any docs with `isAdminUser: false` from the `_users` database to the `tablet_user` database.  We will also have to delete users from this database for any we promote to admin after the promotion.

One snag is that there is still the issue of `localhost` replications failing, even within the same CouchDB.

**Comment by paulbert** [comment]
Fixed in #1486
